test/benchmarks/src/main/scala/scala/ArrayBenchmark.scala

```diff
@@ -0,0 +1,134 @@
+package scala
+
+import java.util.concurrent.TimeUnit
+
+import scala.util.Random
+
+import org.openjdk.jmh.annotations._
+
+@BenchmarkMode(Array(Mode.AverageTime))
+@Fork(2)
+@Threads(1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Benchmark)
+class ArrayBenchmark {
+
+  // From #1338: Optimize simple for loops
+
+  @Param(Array("10", "100", "1000"))
+  var loopSize: Int = _
+  var doubles: Array[Array[Double]] = _
+
+  @Setup(Level.Trial) def initArrays: Unit = {
+    // doubles = Array.ofDim[Double](loopSize,loopSize)
+    // doubles = Array.tabulate(loopSize, loopSize)((i,_) => i.toDouble)
+    doubles = Array.tabulate(loopSize, loopSize) { (_, _) =>
+      scala.math.sqrt(Random.nextDouble)
+    }
+  }
+
+  @Benchmark 
+  def matMulUsingIteratorsBench(): Unit = {
+    matMulUsingIterators(doubles, doubles, doubles)
+  }
+
+  def matMulUsingIterators(
+    a: Array[Array[Double]],
+    b: Array[Array[Double]],
+    c: Array[Array[Double]]
+  ): Unit = {
+
+    val b_j = new Array[Double](b.length)
+
+    for (j <- 0 until b(0).length) {
+      for (k <- 0 until b.length) {
+        b_j(k) = b(k)(j)
+      }
+      for (i <- 0 until a.length) {
+        val c_i = c(i)
+        val a_i = a(i)
+        var s = 0.0d;
+        for (k <- 0 until b.length) {
+          s += a_i(k) * b_j(k)
+        }
+        c_i(j) = s
+      }
+    }
+  }
+
+  @Benchmark 
+  def matMulUsingRangesBench(): Unit = {
+    matMulUsingRanges(doubles, doubles, doubles)
+  }
+
+  def matMulUsingRanges(
```

Comment:
The difference to the benchmark above seems to be the hoisting of the ranges, so fewer range objects are created. Maybe reflect that in the name?

Reply:
Yes, I think that's the only difference.  There is a lot of variable assignment I predict to experiment that it's not the performance issue with the for loop.

There's actually a second version of the code in a second comment, that includes a fourth method, `matMulUsingLimits`.

I copied the original benchmark, blemishes and all, to try and validate the results from ten years ago.  It's probably not worth preserving.